include/ensmallen_bits/utility/indicators/igd.hpp

```diff
@@ -0,0 +1,92 @@
+/**
+ * @file igd.hpp
+ * @author Satyam Shukla
+ *
+ * Inverse Generational Distance Plus (IGD) indicator.
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef ENSMALLEN_INDICATORS_IGD_HPP
+#define ENSMALLEN_INDICATORS_IGD_HPP
+
+namespace ens {
+
+/**
+ * The IGD indicator returns the average distance from each point in the reference
+ * front to the nearest point to it's solution. 
+ *
+ * \f[ d(z,a) = \sqrt{\sum_{i = 1}^{n}(a_i - z_i)^2 \ } \
+ *    \f]
+ *
+ * For more information see:
+ *
+ * @code
+ * @inproceedings{coello2004study,
+ * title={A study of the parallelization of a coevolutionary multi-objective evolutionary algorithm},
+ * author={Coello Coello, Carlos A and Reyes Sierra, Margarita},
+ * booktitle={MICAI 2004: Advances in Artificial Intelligence: Third Mexican International Conference on Artificial Intelligence, Mexico City, Mexico, April 26-30, 2004. Proceedings 3},
+ * pages={688--697},
+ * year={2004},
+ * organization={Springer}
+ * }
+ * @endcode
```

Comment:
It might be nice to add a little documentation mentioning what the purpose of the indicator is (I guess that it is used by AGEMOEA?).  If this is only used by AGEMOEA, I might suggest putting it in the `age_moea/` directory.

Reply:
It is a seperate indicator with a simple change in the already implemented igd_plus score which is used in various papers as the main evaluation metric.

Reply:
I don't see anything in this PR about the "igd_plus score"; can you clarify what you mean?  If you mean that this class is a standalone utility to measure the quality of a multiobjective solution, are we using it anywhere or in any tests?  I don't see it actually used in this PR.

Do you think we can add tests for it?  And, it should be mentioned in the markdown documentation in `doc/` somewhere, because if not, a user will never know it exists.  I might suggest in the `function_types.md` documentation for multi-objective functions.  If we add enough classes of this type, it might make sense to create a separate documentation file for that, but that can happen at a later time.

Lastly I might suggest changing the name `indicators` to something more like `scores` or `metrics` or similar.  I've never seen the work "indicator" used in this way; in fact, you even used the word 'metric' in the description of the class, so maybe that is the best name to use for a directory containing these.

Reply:
Hey, Sure I will add the tests I had done for verification and the reference implementation for the IGD performance indicator with the suggested style changes.

By the above point I actually meant that IGD is very similar to IGD+ which is already implemented in the library and keeping them both together in the indicators should be more better

Regarding the directory name I would suggest to keep the indicator name since these metrics are famously known as `performance indicators `in [literature](https://www.researchgate.net/publication/346979948_Performance_indicators_in_multiobjective_optimization), but yes people do use the name `metrics` as well for it.

Finally for the documentation I will add these performance indicators description in the `function_types.md` file

Reply:
Oh wow, I see now that this is not the first thing in the `indicators/` directory.  I didn't know about the `IGDPlus` class; I see there is an `Epsilon` class too.  (Those are both tested, but not documented.)  If you want to keep the indicator name, that's fine with me (I wasn't familiar with the paper).

When you add the performance indicators description, if you don't mind adding a bit for `IGDPlus` and `Epsilon` too, that would be great; otherwise from a user perspective they effectively don't exist. :)  (Alternately, if you just want to open an issue about it, or do it in a separate PR, that's fine with me, but we should definitely remember to do it at some point!)